[Photo to Product] Reset details from previous image when no text detected

selanthiraiyan opened the pull request:
Closes: #10293 
Closes: #10302 

## Description

When we cannot detect text from a new scan, show an error message and also reset scanned and generated texts from the previous image.

## Testing instructions
- Since the feature is now under A/B test, please enable it by returning `true` in `AddProductFromImageEligibilityChecker`.
- Log in to a WPCom store and select the Products tab.
- Select the "+" button to add a new product.
- Select the manual option if the store has no product.
- Select Simple physical product in the product type action sheet.
- Notice the product details from image form is displayed.
- Select an image with text in it.
- Wait for AI generation to auto till name and description fields. 
- Now select an image without any text in it.
- Observe that you see the message `No text detected. Please select another packaging photo or enter product details manually.` when no text is detected from the image.
- Validate that the scanned and generated texts from the previous image are cleared.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/524475/20a48ca6-fff5-49f0-ad9b-ef1dce42d19f



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Reset scanned texts and generated content from previous image if scan fails.

selanthiraiyan committed:
No longer have to check for `scannedTexts` contents as we are emptying it on error.

selanthiraiyan committed:
Remove unit test checking `scannedTexts` contents.

selanthiraiyan committed:
Fix typo.

selanthiraiyan committed:
Test that text scan failure resets details from previous scan.

selanthiraiyan committed:
Update release notes to ensure that this gets beta tested.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:
```diff
@@ -165,17 +165,17 @@ private extension AddProductFromImageViewModel {
                 [nameViewModel, descriptionViewModel].forEach { $0.reset() }
                 generateProductDetails()
             } catch {
+                // Reset scanned texts and generated content from previous image
+                scannedTexts = []
+                [nameViewModel, descriptionViewModel].forEach { $0.reset() }
+
```
❓ Is there any reason why this is not reset at the beginning of the method instead? I thought we'd want to reset the states when a new image comes in.

selanthiraiyan replied:
Good point. I didn't think of this. Done in https://github.com/woocommerce/woocommerce-ios/pull/10297/commits/8465b078080b639969425de6f31794f20425b277

itsmeichigo reviewed (commented):
Do you think we should reset the state when adding a new image instead of only resetting when an error happens? Or is there some case that I missed?

I also found a case where I opened the image picker and dismissed it - the `onSelectedImage` method is triggered and the product details are generated again:


https://github.com/woocommerce/woocommerce-ios/assets/5533851/4d628c37-49fd-4b82-a2cd-ce9a0956e384

How about adding `removeDuplicates` when observing the image state here?

https://github.com/woocommerce/woocommerce-ios/blob/0eead53984432a9f7afe4ac516890e87352ac055/WooCommerce/Classes/ViewRelated/Products/Add%20Product/AddProductFromImage/AddProductFromImageViewModel.swift#L98-L102

selanthiraiyan committed:
Reset scanned text right after selected an image.

selanthiraiyan committed:
Stop triggering `onSelectedImage` if the image is not changed.

selanthiraiyan committed:
Allow changing result in mock file.

selanthiraiyan committed:
Unit test that new Image selection resets details from previous image.

selanthiraiyan committed:
Update release notes.

selanthiraiyan commented:
> How about adding removeDuplicates when observing the image state here?

I tried adding `removeDuplicates`, but later during testing I found that it creates other edge case issues with `textDetectionErrorMessage` being reset. 

IMO, it might make sense to not alter the `imageState` at all upon dismiss. I logged an issue here https://github.com/woocommerce/woocommerce-ios/issues/10302 to look into it later.

Thanks for the review and testing, @itsmeichigo. The PR is ready for another look.

selanthiraiyan committed:
Remove duplicate method call.

selanthiraiyan committed:
Stop using remove duplicates which adds other edge case errors with `textDetectionErrorMessage` error.

itsmeichigo commented:
> I tried adding removeDuplicates, but later during testing I found that it creates other edge case issues with textDetectionErrorMessage being reset.
IMO, it might make sense to not alter the imageState at all upon dismiss. I logged an issue here https://github.com/woocommerce/woocommerce-ios/issues/10302 to look into it later.

I see that you added `removeDuplicates` on image state, how about adding it to the image instead? 

```Swift
$imageState
    .compactMap { $0.image?.image }
    .removeDuplicates()
```

This seemed to work for me.

selanthiraiyan committed:
Stop calling `onSelectedImage` when image is not changed.

selanthiraiyan committed:
Stop listening to imageState to reset `textDetectionErrorMessage`

selanthiraiyan commented:
> I see that you added removeDuplicates on image state, how about adding it to the image instead? ... This seemed to work for me.

The edge was related to [this code](https://github.com/woocommerce/woocommerce-ios/pull/10297/commits/d1cc7a815aa08f03e5944d9c21ec397707f86ade#diff-ce7c2e92e9359f06b3d0529a06c9f7851052eff122ef909c5dba38ea44238fd9L106-L109) which was listening to imageState to reset `textDetectionErrorMessage`. 

I removed that and reset `textDetectionErrorMessage` value in `onSelectedImage` function. 

- Added `removeDuplicates` in 02b371b43aed12bb3c5abe38555d113210c60feb
- Reset `textDetectionErrorMessage` value in `onSelectedImage` d1cc7a815aa08f03e5944d9c21ec397707f86ade

This works fine from my testing. Kindly let me know your thoughts.

itsmeichigo reviewed (approved):
Thanks for the updates, everything looks good now! 🚢
